spi/imx6ull-ecspi/ecspi-lib.c

```diff
@@ -0,0 +1,303 @@
+	{ { pctl_mux_csi_d3,      3 }, { pctl_mux_csi_d2,     3 }, { pctl_mux_csi_d0,    3 }, { pctl_mux_csi_d1,     3 },
+	  { pctl_mux_lcd_hsync,   8 }, { pctl_mux_lcd_vsync,  8 }, { pctl_mux_lcd_rst,   8 } },
+	{ { pctl_mux_uart2_rts,   8 }, { pctl_mux_uart2_cts,  8 }, { pctl_mux_uart2_rx,  8 }, { pctl_mux_uart2_tx,   8 },
+	  { pctl_mux_nand_ale,    8 }, { pctl_mux_nand_re,    8 }, { pctl_mux_nand_we,   8 } },
+	{ { pctl_mux_enet2_txclk, 3 }, { pctl_mux_enet2_txen, 3 }, { pctl_mux_enet2_tx1, 3 }, { pctl_mux_enet2_rxer, 3 },
+	  { pctl_mux_nand_d1,     8 }, { pctl_mux_nand_d2,    8 }, { pctl_mux_nand_d3,   8 } }
+};
+
+ecspi_pctl_t ecspi_pctl_isel[4][4] = {
+	{ { pctl_isel_ecspi1_miso, 1 }, { pctl_isel_ecspi1_mosi, 1 }, { pctl_isel_ecspi1_sclk, 1 }, { pctl_isel_ecspi1_ss0, 1 } },
+	{ { pctl_isel_ecspi2_miso, 0 }, { pctl_isel_ecspi2_mosi, 1 }, { pctl_isel_ecspi2_sclk, 0 }, { pctl_isel_ecspi2_ss0, 0 } },
+	{ { pctl_isel_ecspi3_miso, 0 }, { pctl_isel_ecspi3_mosi, 0 }, { pctl_isel_ecspi3_sclk, 0 }, { pctl_isel_ecspi3_ss0, 0 } },
+	{ { pctl_isel_ecspi4_miso, 0 }, { pctl_isel_ecspi4_mosi, 0 }, { pctl_isel_ecspi4_sclk, 0 }, { pctl_isel_ecspi4_ss0, 0 } }
+};
+
+uint32_t ecspi_pctl_clk[4] = { pctl_clk_ecspi1, pctl_clk_ecspi2, pctl_clk_ecspi3, pctl_clk_ecspi4 };
+
+static ecspi_t ecspi[4] = {0};
+
+
+static void ecspi_setBurst(int dev_no, uint16_t burst)
+{
+	ecspi_t *e;
+
+	if (burst % 32 == 1) {
+		printf("ecspi: burst length of 32n + 1 is erroneous in hardware itself\n");
+		return;
+	}
+
+	e = &ecspi[dev_no - 1];
+	*(e->base + conreg) = (*(e->base + conreg) & ~(0xFFF << 20)) | ((burst - 1) << 20);
+}
+
+static void set_mux(int dev_no, uint8_t chan_msk)
+{
+	platformctl_t ctl;
+	int i;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_iomux;
+
+	ctl.iomux.sion = 0;
+
+	for (i = 0; i < 7; i++) {
+		/* Skip SS lines not enabled in chan_msk */
+		if (i < 3 || (chan_msk & (1 << (i - 3)))) {
+			ctl.iomux.mux = ecspi_pctl_mux[(dev_no - 1)][i].pctl;
+			ctl.iomux.mode = ecspi_pctl_mux[(dev_no - 1)][i].val;
+			platformctl(&ctl);
+		}
+	}
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_ioisel;
+
+	for (i = 0; i < 3; i++) {
+		if (i < 3 || (chan_msk & (1 << (i - 3)))) {
+			ctl.ioisel.isel = ecspi_pctl_isel[(dev_no - 1)][i].pctl;
+			ctl.ioisel.daisy = ecspi_pctl_isel[(dev_no - 1)][i].val;
+			platformctl(&ctl);
+		}
+	}
+}
+
+static void set_clk(int dev_no)
+{
+	platformctl_t ctl;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_devclock;
+	ctl.devclock.dev = ecspi_pctl_clk[dev_no - 1];
+	ctl.devclock.state = 0b11;
+
+	platformctl(&ctl);
+}
+
+static void ecspi_writeFifo(int dev_no, const uint8_t *out, uint32_t len)
+{
+	uint32_t word;
+	ecspi_t *e = &ecspi[dev_no - 1];
+	uint8_t fill_len = len % 4;
+
+	if (fill_len > 0) {
+		word = 0;
+
+		while (fill_len > 0) {
+			word = (word << 8) | *out++;
+			fill_len--;
+			len--;
+		}
+
+		*(e->base + txdata) = word;
+	}
+
+	while (len > 0) {
+		word = out[3] | (out[2] << 8) | (out[1] << 16) | (out[0] << 24);
```

Comment:
Maybe cast _out_ to uint32_t?

Reply:
Casting will reorder bytes which is unacceptable.